sdks/java/io/google-cloud-platform/src/main/java/org/apache/beam/sdk/io/gcp/bigtable/BigtableServiceImpl.java

```diff
@@ -210,6 +212,254 @@ public Row getCurrentRow() throws NoSuchElementException {
+    }
+
+    @Override
+    public boolean start() throws IOException {
+      future = fetchNextSegment();
+      return advance();
+    }
+
+    @Override
+    public boolean advance() throws IOException {
+      if (buffer.size() < refillSegmentWaterMark && future == null) {
+        future = fetchNextSegment();
+      }
+      if (buffer.isEmpty() && future != null) {
+        waitReadRowsFuture();
+      }
+      // The last requst will return an empty rowSet which will mean that there are no more rows to read
+      if (buffer.isEmpty()) {
+        return false;
+      }
+      currentRow = buffer.remove();
+      return currentRow != null;
+    }
+
+    Future<UpstreamResults> fetchNextSegment() {
+      SettableFuture<UpstreamResults> f = SettableFuture.create();
+      if (nextRequest == null) {
+        f.set(new UpstreamResults(ImmutableList.of(), null));
+        return f;
+      }
+
+      // TODO(diegomez): Remove atomic ScanHandler for simpler StreamObserver/Future implementation
+      AtomicReference<ScanHandler> atomic = new AtomicReference<>();
+      ScanHandler handler =
+          session
+              .getDataClient()
+              .readFlatRows(
+                  nextRequest,
+                  new StreamObserver<FlatRow>() {
+                    List<Row> rows = new ArrayList<>();
+                    long currentByteSize = 0;
+                    boolean byteLimitReached = false;
+
+                    @Override
+                    public void onNext(FlatRow flatRow) {
+                      currentByteSize += computeRowSize(flatRow);
+                      Row row = FlatRowConverter.convert(flatRow);
+                      rows.add(row);
+
+                      if (currentByteSize > bufferByteLimit) {
+                        byteLimitReached = true;
+                        atomic.get().cancel();
+                        return;
+                      }
+                    }
+
+                    @Override
+                    public void onError(Throwable e) {
+                      f.setException(e);
+                    }
+
+                    @Override
+                    public void onCompleted() {
+                      ReadRowsRequest nextNextRequest = null;
+
+                      if (byteLimitReached || rows.size() == segmentSize) {
+                        nextNextRequest =
+                            truncateRequest(nextRequest, rows.get(rows.size() - 1).getKey());
+                      }
+
+                      f.set(new UpstreamResults(rows, nextNextRequest));
+                    }
+                  });
+      atomic.set(handler);
+      return f;
+    }
+
+    void waitReadRowsFuture() throws IOException {
+      try {
+        UpstreamResults r = future.get();
+        buffer.addAll(r.rows);
+        nextRequest = r.nextRequest;
+        future = null;
+        serviceCallMetric.call("ok");
+      } catch (StatusRuntimeException e) {
```

Comment:
You wont get a StatusRuntimeException here. It will be wrapped in an ExecutionException. So you need to check the cause in the catch below.
Also please have a test to cover this